pkg/openstack/nova.go

```diff
@@ -45,9 +51,110 @@ func ReconcileNova(ctx context.Context, instance *corev1beta1.OpenStackControlPl
+		}
+		instance.Spec.Nova.Template.APIServiceTemplate.Override.Service[string(endpointType)] =
+			AddServiceComponentLabel(
+				ptr.To(instance.Spec.Nova.Template.APIServiceTemplate.Override.Service[string(endpointType)]),
+				nova.Name+"-api")
+
+		// cell NoVNCProxy service override
+		for cellName := range instance.Spec.Nova.Template.CellTemplates {
+			cellTemplate := instance.Spec.Nova.Template.CellTemplates[cellName]
+			if cellTemplate.NoVNCProxyServiceTemplate.Override.Service == nil {
+				cellTemplate.NoVNCProxyServiceTemplate.Override.Service = map[string]service.RoutedOverrideSpec{}
+			}
+			cellTemplate.NoVNCProxyServiceTemplate.Override.Service[string(endpointType)] =
+				AddServiceComponentLabel(
+					ptr.To(cellTemplate.NoVNCProxyServiceTemplate.Override.Service[string(endpointType)]),
+					getNoVNCProxyServiceLabel(nova.Name, cellName))
+			instance.Spec.Nova.Template.CellTemplates[cellName] = cellTemplate
+		}
+	}
+
+	// When component services got created check if there is the need to create a route
+	if err := helper.GetClient().Get(ctx, types.NamespacedName{Name: "nova", Namespace: instance.Namespace}, nova); err != nil {
+		if !k8s_errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Nova API
+	if nova.Status.Conditions.IsTrue(novav1.NovaAPIReadyCondition) {
+		svcs, err := service.GetServicesListWithLabel(
+			ctx,
+			helper,
+			instance.Namespace,
+			map[string]string{common.AppSelector: nova.Name + "-api"},
+		)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		var ctrlResult reconcile.Result
+		instance.Spec.Nova.Template.APIServiceTemplate.Override.Service, ctrlResult, err = EnsureRoute(
+			ctx,
+			instance,
+			helper,
+			nova,
+			svcs,
+			instance.Spec.Nova.Template.APIServiceTemplate.Override.Service,
+			instance.Spec.Nova.APIOverride.Route,
+			corev1beta1.OpenStackControlPlaneExposeNovaReadyCondition,
+		)
+		if err != nil {
+			return ctrlResult, err
+		} else if (ctrlResult != ctrl.Result{}) {
+			return ctrlResult, nil
+		}
+	}
+
+	if nova.Status.Conditions.IsTrue(novav1.NovaAllCellsDBReadyCondition) {
+		// cell NoVNCProxy
+		for cellName := range instance.Spec.Nova.Template.CellTemplates {
+			labelValue := getNoVNCProxyServiceLabel(nova.Name, cellName)
+			cellTemplate := instance.Spec.Nova.Template.CellTemplates[cellName]
+
+			svcs, err := service.GetServicesListWithLabel(
+				ctx,
+				helper,
+				instance.Namespace,
+				map[string]string{common.AppSelector: labelValue},
+			)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			var ctrlResult reconcile.Result
+			cellTemplate.NoVNCProxyServiceTemplate.Override.Service, ctrlResult, err = EnsureRoute(
+				ctx,
+				instance,
+				helper,
+				nova,
+				svcs,
+				cellTemplate.NoVNCProxyServiceTemplate.Override.Service,
+				instance.Spec.Nova.CellOverride[cellName].NoVNCProxy.Route,
+				corev1beta1.OpenStackControlPlaneExposeNovaReadyCondition,
```

Comment:
This is problematic. There are multiple EnsureRoute calls for a single reconcile and each of that call sets the same condition. This means that if there are multiple errors then the earlier errors are lost. Also if there is an error then a success then the success will override the error. E.g. if the route creation of the NovaAPI fails but the the route creation of the NoVNCProxy succeeds the later will set the condition to True and we loose the error.

Reply:
Sorry I mixed up. You return on each error. So the only overlap is when a previous success is overridden by a later failure. But that I guess is OK. The above described problem only arises when we switch to parallel route creation. But we don't need that now. So you can ignore the above comment

Reply:
right, I think its ok to return on the first failure.